Start and stop the Gatekeeper status sync controller dynamically

mprahl opened the pull request:
Rather than rely on a health probe failing if the Gatekeeper installation status changes, just dynamically start and stop a new manager.

Relates:
https://issues.redhat.com/browse/ACM-10966

JustinKuli commented on main.go:
```diff
@@ -951,3 +903,252 @@ func addControllers(
+	mgrOptions.LeaderElectionID = "governance-policy-framework-addon3.open-cluster-management.io"
+	mgrOptions.HealthProbeBindAddress = healthAddress
+
+	startMgr := func(ctx context.Context) {
```
I think this could be defined outside of the `gatekeeperInstallationChecker` function. It might need more parameters, but I think it would help readability a lot.

JustinKuli commented on main.go:
```diff
@@ -951,3 +903,252 @@ func addControllers(
+				select {
+				case <-ctx.Done():
+					// Stop the retry watcher if the parent context is canceled.
+					w.Stop()
```
Since this is inside the `<- w.Done()` case, isn't the watcher already stopped?

mprahl replied:
I'm not certain about this. I couldn't find any handling of a closed context in the client-go code but it's not harmful to include this since it explicitly tells the retry loop to stop.

JustinKuli commented on main.go:
```diff
@@ -951,3 +900,259 @@ func addControllers(
+
+	mgrRunning := false
+
+	if gatekeeperInstalled {
+		mgrRunning = true
+
+		mgrCtx, mgrCtxCancel = context.WithCancel(ctx)
+
+		go startMgr(mgrCtx)
+	}
```
Is it possible to adjust the case in the select below to do this "initial" start there? My feeling is that "recover from stop" should be the same as "initial start".

JustinKuli commented on main.go:
```diff
@@ -951,3 +900,262 @@ func addControllers(
+// installation status changes. The controller must be off when Gatekeeper is not installed.
+func gatekeeperInstallationChecker(
+	ctx context.Context, managedCfg *rest.Config, dynamicClient dynamic.Interface, mgrOptions manager.Options,
+) error {
```
I think the cases where this function returns an error are just:
1. if it can't get a free local address
2. if the initial attempt to make the retry watcher fails.

So, if case 1 moves out of the function, and case 2 could be handled like non-initial retry watcher creation failures are handled (by retrying, I think?) then this function would not return an error.

The reason I'm thinking in that direction is that this function ends with a `go func() ...`, so I feel like most of its "work" is asynchronous. But the name of the function, and the calling convention doesn't make that obvious. If we called it as `go gkInstallChecker(...)`, it might be more clear what it is doing? Plus it would save some indents at the end.

I don't know - maybe it wouldn't be better. Just something to consider

mprahl commented on main.go:
```diff
@@ -275,12 +281,13 @@ func main() {
 
-	healthAddresses := []string{mgrHealthAddr}
+	healthAddressesLock.RLock()
+	healthAddresses[mgrHealthAddr] = true
```
Good catch. Thanks!

mprahl commented:
@JustinKuli could you please take another look?

mprahl commented on main.go:
```diff
@@ -291,11 +298,36 @@ func main() {
+		if semver.Compare(serverVersion.GitVersion, "v1.16.0") >= 0 {
+			dynamicClient := dynamic.NewForConfigOrDie(managedCfg)
+
+			go manageGatekeeperSyncManager(mgrCtx, managedCfg, dynamicClient, mgrOptionsBase)
```
Yeah, I think so!

mprahl commented:
> Maybe it's just had more time to settle in my head, but I think this is looking great

Thanks @JustinKuli! I addressed your comments and I had forgotten about the use of the wait group in the main function, so I took that into account when the Gatekeeper sync manager is running.

JustinKuli reviewed (approved):
LGTM
